AlfrescoApp/Model/Managers/RealmSyncManager+Refresh.m

```diff
@@ -0,0 +1,217 @@
+        // STEP 5 - Start downloading any new content that appears inside the sync set
+        [self startDownloadingNewContent];
+        
+        if (completionBlock)
+        {
+            completionBlock(YES);
+        }
+    }];
+}
+
+#pragma mark - Private Methods
+
+- (void)markTopLevelNodesAsPending
+{
+    RLMRealm *realm = self.mainThreadRealm;
+    // Get all top level nodes.
+    RLMResults *allTopLevelNodes = [[RealmManager sharedManager] topLevelNodesInRealm:realm];
+    
+    // Mark all top level nodes as "pending sync" status.
+    for (RealmSyncNodeInfo *node in allTopLevelNodes)
+    {
+        SyncNodeStatus *nodeStatus = [[RealmSyncManager sharedManager] syncStatusForNodeWithId:node.syncNodeInfoId];
+        nodeStatus.status = SyncStatusWaiting;
+    }
+}
+
+- (void)invalidateParentChildHierarchy
+{
+    RLMRealm *realm = self.mainThreadRealm;
+    RLMResults *allNodes = [[RealmManager sharedManager] allNodesInRealm:realm];
+    
+    for (RealmSyncNodeInfo *node in allNodes)
+    {
+        if (node.parentNode)
+        {
+            [realm beginWriteTransaction];
+            node.parentNode = nil;
+            [realm commitWriteTransaction];
+        }
+    }
+}
+
+- (void)rebuildDataBaseWithCompletionBlock:(void (^)(BOOL completed))completionBlock
+{
+    RLMRealm *realm = self.mainThreadRealm;
+    RLMResults *topLevelFolders = [[RealmManager sharedManager] topLevelFoldersInRealm:realm];
+    
+    self.nodeChildrenRequestsCount = 0;
+    
+    for (RealmSyncNodeInfo *node in topLevelFolders)
+    {
+        AlfrescoFolder *folder = (AlfrescoFolder *)node.alfrescoNode;
+        
+        [self retrieveNodeHierarchyForFolder:folder withCompletionBlock:^(BOOL completed){
+            if(self.nodeChildrenRequestsCount == 0)
+            {
+                if(completionBlock)
+                {
+                    completionBlock(completed);
+                }
+            }
+        }];
+    }
+}
+
+- (void)retrieveNodeHierarchyForFolder:(AlfrescoFolder *)folder withCompletionBlock:(void (^)(BOOL completed))completionBlock
+{
+    self.nodeChildrenRequestsCount++;
+    
+    [self.documentFolderService retrieveChildrenInFolder:folder completionBlock:^(NSArray *array, NSError *error){
+        self.nodeChildrenRequestsCount--;
+        
+        if (array)
+        {
+            for (AlfrescoNode *childNode in array)
+            {
+                [self updateDataBaseForChildNode:childNode withParent:folder];
+                
+                if (childNode.isFolder)
+                {
+                    [self retrieveNodeHierarchyForFolder:(AlfrescoFolder *)childNode withCompletionBlock:^(BOOL completed){
+                        if (completionBlock != NULL)
+                        {
+                            completionBlock(YES);
+                        }
+                    }];
+                }
+            }
+        }
+        else
+        {
+            RLMRealm *realm = self.mainThreadRealm;
```

Comment:
Please use the defaultRealm method to get an instance of a RLMRealm object instead of the mainThreadRealm in order to ensure that any object obtained is accessed from the correct thread.